app/models/missing_link_finder.rb

```diff
@@ -7,58 +15,78 @@ class MissingLinkFinder
 
   def initialize(redirection)
     @redirection = redirection
+    @session = Capybara::Session.new(:mechanize)
   end
 
   def run
-    begin
-      response = fetch(redirection.url)
-      if response.is_a?(Net::HTTPSuccess)
-        missing = missing_links(response.body)
-        if missing.empty?
-          { status: :good }
-        else
-          { status: :missing_links, missing: missing }
-        end
+    session.visit(redirection.url)
+    if session.status_code.in?(200..299)
+      if missing_links.empty?
+        {status: :good}
       else
-        { status: :offline }
+        {status: :missing_links, missing: missing_links}
       end
-    rescue SocketError
-      { status: :offline }
-    rescue StandardError => error
-      { status: :error, error: "#{error.class}: #{error.message}" }
+    else
+      {status: :offline}
+    end
+  rescue SocketError
+    {status: :offline}
+  rescue => error
+    if error.message.include?("ERR_NAME_NOT_RESOLVED")
+      {status: :offline}
+    elsif error.message.include?("Net::HTTPNotFound")
+      {status: :not_found}
+    else
+      {status: :error, error: "#{error.class}: #{error.message}"}
     end
   end
 
   private
 
-  attr_reader :redirection
+  attr_reader :redirection, :session
 
-  def missing_links(html)
-    if html
-      missing_links = []
-      missing_links << "next" if !html.match?(next_link)
-      missing_links << "prev" if !html.match?(prev_link)
+  def missing_links
+    missing_next = !has_next_link?
+    missing_previous = !has_prev_link?
+    session.all("iframe, frame").map { |i| i[:src] }.each do |iframe_src|
+      if missing_next || missing_previous
+        session.visit ensure_absolute_path(iframe_src)
+        if has_next_link?
```

Comment:
I elected to use `if` statements instead of the following code, which I think is some of the most confusing I've seen in years 😄 

```ruby
missing_next &&= !has_next_link?
missing_previous &&= !has_prev_link?
```

Reply:
Wow. I did not know that was even valid ruby. Good call.